[5.1][String] Scalar-alignment bug fixes.

milseman opened the pull request:
5.1 cherry-pick of https://github.com/apple/swift/pull/23834

Explanation: All String views use the same index type, with the semantics being that we interpret them as representing the nearest prior scalar-aligned position. However, we were not checking or aligning indices previously, leading to crashes in the (somewhat uncommon) situation that a non-scalar-aligned index was interchanged between sub-scalar views. This change checks and performs alignment, and allocates a known-to-be-scalar-aligned bit in String.Index’s ABI to recoup the performance back.

Scope: Swift Standard Library

Issue: <rdar://problem/51199363> and https://bugs.swift.org/browse/SR-10124

Risk: Low. There is a chance that code that was improperly exchanging indices between different Strings (not just different views) will not get the correct behavior provided by this fix, if the known-to-be-aligned bit were set from one String.

Testing: Full integration testing, benchmarking

Reviewer: @stephentyrone, @lorentey, @Catfish-Man

milseman committed:
[String] Slice contents before asking ICU

ICU will return different results if we call with an offset into a
code unit buffer vs if we slice the buffer first and provide an offset
of zero. Slicing more closely models the semantics of SE-0180, so use
that.

Test case coming in subsequent commit enforcing index
scalar-alignment.

milseman committed:
[gardening] Remove trailing whitespace

milseman committed:
[String] Scalar-alignment bug fixes.

Fixes a general category (pun intended) of scalar-alignment bugs
surrounding exchanging non-scalar-aligned indices between views and
for slicing.

SE-0180 unifies the Index type of String and all its views and allows
non-scalar-aligned indices to be used across views. In order to
guarantee behavior, we often have to check and perform scalar
alignment. To speed up these checks, we allocate a bit denoting
known-to-be-aligned, so that the alignment check can skip the
load. The below shows what views need to check for alignment before
they can operate, and whether the indices they produce are aligned.

┌───────────────╥────────────────────┬──────────────────────────┐
│ View          ║ Requires Alignment │ Produces Aligned Indices │
╞═══════════════╬════════════════════╪══════════════════════════╡
│ Native UTF8   ║ no                 │ no                       │
├───────────────╫────────────────────┼──────────────────────────┤
│ Native UTF16  ║ yes                │ no                       │
╞═══════════════╬════════════════════╪══════════════════════════╡
│ Foreign UTF8  ║ yes                │ no                       │
├───────────────╫────────────────────┼──────────────────────────┤
│ Foreign UTF16 ║ no                 │ no                       │
╞═══════════════╬════════════════════╪══════════════════════════╡
│ UnicodeScalar ║ yes                │ yes                      │
├───────────────╫────────────────────┼──────────────────────────┤
│ Character     ║ yes                │ yes                      │
└───────────────╨────────────────────┴──────────────────────────┘

The "requires alignment" applies to any operation taking a
String.Index that's not defined entirely in terms of other operations
taking a String.Index. These include:

* index(after:)
* index(before:)
* subscript
* distance(from:to:) (since `to` is compared against directly)
* UTF16View._nativeGetOffset(for:)

milseman committed:
[gardening] Add underscore to internal member

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test os x platform

lorentey commented:
```
11:00:17 Test Suite 'Selected tests' started at 2019-06-28 11:00:03.384
11:00:17 Test Suite 'SwiftPMPackageTests.xctest' started at 2019-06-28 11:00:03.385
11:00:17 Test Suite 'PackageDescription4_2LoadingTests' started at 2019-06-28 11:00:03.385
11:00:17 Test Case '-[PackageLoadingTests.PackageDescription4_2LoadingTests testDuplicateDependencyDecl]' started.
11:00:17 <unknown>:0: error: -[PackageLoadingTests.PackageDescription4_2LoadingTests testDuplicateDependencyDecl] : failed: caught error: The operation couldn’t be completed. (PackageLoading.ManifestParseError error 0.)
11:00:17 Test Case '-[PackageLoadingTests.PackageDescription4_2LoadingTests testDuplicateDependencyDecl]' failed (0.435 seconds).
11:00:17 Test Suite 'PackageDescription4_2LoadingTests' failed at 2019-06-28 11:00:03.820.
11:00:17 	 Executed 1 test, with 1 failure (1 unexpected) in 0.435 (0.436) seconds
11:00:17 Test Suite 'SwiftPMPackageTests.xctest' failed at 2019-06-28 11:00:03.820.
11:00:17 	 Executed 1 test, with 1 failure (1 unexpected) in 0.435 (0.436) seconds
11:00:17 Test Suite 'Selected tests' failed at 2019-06-28 11:00:03.820.
11:00:17 	 Executed 1 test, with 1 failure (1 unexpected) in 0.435 (0.436) seconds
```

lorentey commented:
@swift-ci test macOS platform

lorentey commented:
🤔
```
15:13:03 ("Stack dump:\n0.\tProgram arguments: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swift-macosx-x86_64/bin/swift -frontend -interpret /var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/TemporaryFile.9qwtEz.swift -target x86_64-apple-macosx10.10 -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -I /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swiftpm-macosx-x86_64/x86_64-apple-macosx/lib/swift/pm/4_2 -swift-version 5 -package-description-version 5 -module-name main -lPackageDescription -- -fileno 42 \n0  swift                       0x000000010e6fefe5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37\n1  swift                       0x000000010e6fe028 llvm::sys::RunSignalHandlers() + 248\n2  swift                       0x000000010e6ff5d8 SignalHandler(int) + 264\n3  libsystem_platform.dylib    0x00007fff63db6b5d _sigtramp + 29\n4  libsystem_platform.dylib    0x00007ff56de43bc0 _sigtramp + 168349824\n5  libPackageDescription.dylib 0x0000000115e39bb1 $s18PackageDescription14AppleOSVersion33_EA5579823E05354E291299F0EC3450AALLPAAE6stringxSS_tcfCAA17SupportedPlatformV03MacD0V_Tg5Tf4xd_n + 161\n6  libPackageDescription.dylib 0x0000000115e33cf2 globalinit_33_EA5579823E05354E291299F0EC3450AA_func10 + 82\n7  libdispatch.dylib           0x00007fff63b7e63d _dispatch_client_callout + 8\n8  libdispatch.dylib           0x00007fff63b7fd4c _dispatch_once_callout + 20\n9  libswiftCore.dylib          0x00007fff636047e9 swift_once + 25\n10 libPackageDescription.dylib 0x0000000115e33d31 $s18PackageDescription17SupportedPlatformV12MacOSVersionV6v10_13AEvau + 33\n11 libPackageDescription.dylib 0x0000000115eb5079 $s18PackageDescription17SupportedPlatformV12MacOSVersionV6v10_13AEvau + 529257\n12 swift                       0x000000010ab98258 llvm::MCJIT::runFunction(llvm::Function*, llvm::ArrayRef<llvm::GenericValue>) + 456\n13 swift                       0x000000010ab9b3c1 llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, char const* const*) + 1313\n14 swift                       0x000000010ab8daa1 swift::RunImmediately(swift::CompilerInstance&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, swift::IRGenOptions&, swift::SILOptions const&) + 3505\n15 swift                       0x000000010ab7e075 performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 13653\n16 swift                       0x000000010ab79b52 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2978\n17 swift                       0x000000010ab22a28 main + 696\n18 libdyld.dylib               0x00007fff63bcb3d5 start + 1", diagnosticFile: nil)
15:13:03 Test Suite 'Selected tests' started at 2019-06-28 15:12:46.883
15:13:03 Test Suite 'SwiftPMPackageTests.xctest' started at 2019-06-28 15:12:46.883
15:13:03 Test Suite 'PackageDescription5LoadingTests' started at 2019-06-28 15:12:46.884
15:13:03 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' started.
15:13:03 /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swiftpm/Tests/PackageLoadingTests/PD5LoadingTests.swift:166: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed
15:13:03 <unknown>:0: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed: caught error: The operation couldn’t be completed. (PackageLoading.ManifestParseError error 0.)
15:13:03 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' failed (7.880 seconds).
15:13:03 Test Suite 'PackageDescription5LoadingTests' failed at 2019-06-28 15:12:54.764.
15:13:03 	 Executed 1 test, with 2 failures (1 unexpected) in 7.880 (7.880) seconds
15:13:03 Test Suite 'SwiftPMPackageTests.xctest' failed at 2019-06-28 15:12:54.764.
15:13:03 	 Executed 1 test, with 2 failures (1 unexpected) in 7.880 (7.880) seconds
15:13:03 Test Suite 'Selected tests' failed at 2019-06-28 15:12:54.764.
15:13:03 	 Executed 1 test, with 2 failures (1 unexpected) in 7.880 (7.881) seconds
```

lorentey commented:
@swift-ci test macOS platform

lorentey commented:
```
02:02:45 Test Suite 'SwiftPMPackageTests.xctest' started at 2019-06-29 02:02:32.009
02:02:45 Test Suite 'PackageDescription5LoadingTests' started at 2019-06-29 02:02:32.009
02:02:45 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' started.
02:02:45 /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swiftpm/Tests/PackageLoadingTests/PD5LoadingTests.swift:166: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed
02:02:45 <unknown>:0: error: -[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms] : failed: caught error: The operation couldn’t be completed. (PackageLoading.ManifestParseError error 0.)
02:02:45 Test Case '-[PackageLoadingTests.PackageDescription5LoadingTests testPlatforms]' failed (6.399 seconds).
02:02:45 Test Suite 'PackageDescription5LoadingTests' failed at 2019-06-29 02:02:38.408.
02:02:45 	 Executed 1 test, with 2 failures (1 unexpected) in 6.399 (6.400) seconds
02:02:45 Test Suite 'SwiftPMPackageTests.xctest' failed at 2019-06-29 02:02:38.408.
02:02:45 	 Executed 1 test, with 2 failures (1 unexpected) in 6.399 (6.400) seconds
02:02:45 Test Suite 'Selected tests' failed at 2019-06-29 02:02:38.408.
02:02:45 	 Executed 1 test, with 2 failures (1 unexpected) in 6.399 (6.400) seconds
```

milseman committed:
[String] Switch scalar-aligned bit to a reserved bit.

Since scalar-alignment is set in inlinable code, switch the alignment
bit to one of the previously-reserved bits rather than a grapheme
cache bit. Setting a grapheme cache bit in inlinable would break
backward deployment, as older versions would interpret it as a cached
value.

Also adjust the name to "scalar-aligned", which is clearer, and
removed assertion (which should be a real precondition).

milseman commented:
Also cherry-picking https://github.com/apple/swift/pull/25948

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test

stephentyrone commented on stdlib/public/core/Substring.swift:
```diff
@@ -97,9 +97,15 @@ public struct Substring {
   @usableFromInline
   internal var _slice: Slice<String>
 
-  @inlinable @inline(__always)
```
Expected?

milseman replied:
Yup, this is no longer trivial now and we don't want to force the optimizer's hand. The optimizer's probably better off not inlining this call and inlining the `scalarAlign` calls.

milseman commented:
@swift-ci please nominate

stephentyrone commented on stdlib/public/core/UnicodeHelpers.swift:
```diff
@@ -140,10 +140,12 @@ internal func _continuationPayload(_ x: UInt8) -> UInt32 {
   return UInt32(x & 0x3F)
 }
 
-@inlinable @inline(__always)
```
Also this one, to double-check?

milseman replied:
Yeah, the inline-me fast-path has switched to the bit-check.

stephentyrone commented on test/stdlib/StringIndex.swift:
```diff
@@ -232,21 +227,238 @@ StringIndexTests.test("String.Index(_:within) / Range<String.Index>(_:in:)") {
+    file: String = #file,
+    line: UInt = #line
+  ) {
+    guard #available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) else {
```
Can this be macOS 10.15, etc now?

milseman replied:
I'm matching what's currently there. I had also tried to update all of String's versions in another commit, but hit a snag during testing (don't recall what it was, probably I wasn't running the beta). I want to get this in for 5.1 in time, so I just followed what we already have throughout the source.

milseman commented:
@swift-ci nominate

stephentyrone reviewed (approved):
Generally LGTM. Flagged a couple changes to make sure they're intended (or more specifically, that this hasn't picked up some other commits).

Catfish-Man reviewed (approved):
There's a lot here to digest and I'd need to step through some of it in a debugger to be thorough, but reading through it it looks fine.
